polly/lib/Analysis/ScopBuilder.cpp

```diff
@@ -2568,49 +2579,196 @@ bool checkCandidatePairAccesses(MemoryAccess *LoadMA, MemoryAccess *StoreMA,
+      if (auto *ValueInst = dyn_cast<Instruction>(Store->getValueOperand()))
+        State.insert(std::make_pair(Store, State[ValueInst]));
       continue;
+    }
+
+    // Non load and store instructions are either binary operators or they will
+    // invalidate all used loads.
+    auto *BinOp = dyn_cast<BinaryOperator>(&Inst);
+    auto CurRedType = getReductionType(BinOp);
+    LLVM_DEBUG(dbgs() << "CurInst: " << Inst << " RT: " << CurRedType << "\n");
+
+    // Iterate over all operands and propagate their input loads to instruction.
+    FlowInSetTy &InstInFlowSet = State[&Inst];
+    for (Use &Op : Inst.operands()) {
+      auto *OpInst = dyn_cast<Instruction>(Op);
+      if (!OpInst)
+        continue;
+
+      LLVM_DEBUG(dbgs().indent(4) << "Op Inst: " << *OpInst << "\n");
+      const StateTy::iterator &OpInFlowSetIt = State.find(OpInst);
+      if (OpInFlowSetIt == State.end())
+        continue;
+
+      // Iterate over all the input loads of the operand and combine them
+      // with the input loads of current instruction.
+      FlowInSetTy &OpInFlowSet = OpInFlowSetIt->second;
+      for (auto &OpInFlowPair : OpInFlowSet) {
+        unsigned OpFlowIn = OpInFlowPair.second.first;
+        unsigned InstFlowIn = InstInFlowSet[OpInFlowPair.first].first;
+
+        auto OpRedType = OpInFlowPair.second.second;
+        auto InstRedType = InstInFlowSet[OpInFlowPair.first].second;
+
+        auto NewRedType = combineReductionType(OpRedType, CurRedType);
+        if (InstFlowIn)
+          NewRedType = combineReductionType(NewRedType, InstRedType);
+
+        LLVM_DEBUG(dbgs().indent(8) << "OpRedType: " << OpRedType << "\n");
+        LLVM_DEBUG(dbgs().indent(8) << "NewRedType: " << NewRedType << "\n");
+        InstInFlowSet[OpInFlowPair.first] =
+            std::make_pair(OpFlowIn + InstFlowIn, NewRedType);
+      }
+    }
+
+    // If this operation is used outside the block, invalidate all the loads
+    // which feed into it.
+    if (UsedOutsideBlock)
+      for (const auto &FlowInSetElem : InstInFlowSet)
+        InvalidLoads.insert(FlowInSetElem.first);
+  }
+
+  // All used loads are propagated through the whole basic block; now try to
+  // find valid reduction like candidate pairs. These load-store pairs fulfill
+  // all reduction like properties with regards to only this load-store chain.
+  // We later have to check if the loaded value was invalidated by an
+  // instruction not in that chain.
+  using MemAccPair = std::pair<MemoryAccess *, MemoryAccess *>;
+  DenseMap<MemAccPair, MemoryAccess::ReductionType> ValidCandidates;
+  DominatorTree *DT = Stmt.getParent()->getDT();
+
+  // Iterate over all write memory accesses and check the loads flowing into
+  // it for reduction candidate pairs.
+  for (MemoryAccess *WriteMA : Stmt.MemAccs) {
+    if (WriteMA->isRead())
+      continue;
+    StoreInst *St = dyn_cast<StoreInst>(WriteMA->getAccessInstruction());
+    if (!St || St->isVolatile())
```

Comment:
Shouldn't volatile stores lead to invalidating the loads as other non-valid reductions? ScopDetection should already reject volatile access anyway, so this could be a `assert(!St->isVolatile())` or `assert(St->isSimple())`